From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrew Pinski <pinskia@gcc.gnu.org>
Date: Tue, 11 Jun 2024 19:17:04 +0000
Subject: [PATCH 14/14] Avoid <memory> poisoning on musl

On musl <pthread.h> uses calloc() (via <sched.h>). <sstream> includes
it indirectly and exposes use of poisoned calloc() when module code is
built:

	/usr/src/gcc-14.1.0.bld/./prev-gcc/xg++ -B/usr/src/gcc-14.1.0.bld/./prev-gcc/ -B/usr/x86_64-pc-linux-musl/bin/ -nostdinc++ -B/usr/src/gcc-14.1.0.bld/prev-x86_64-pc-linux-musl/libstdc++-v3/src/.libs -B/usr/src/gcc-14.1.0.bld/prev-x86_64-pc-linux-musl/libstdc++-v3/libsupc++/.libs  -isystem /usr/src/gcc-14.1.0.bld/prev-x86_64-pc-linux-musl/libstdc++-v3/include/x86_64-pc-linux-musl  -isystem /usr/src/gcc-14.1.0.bld/prev-x86_64-pc-linux-musl/libstdc++-v3/include  -isystem /usr/src/gcc-14.1.0/libstdc++-v3/libsupc++ -L/usr/src/gcc-14.1.0.bld/prev-x86_64-pc-linux-musl/libstdc++-v3/src/.libs -L/usr/src/gcc-14.1.0.bld/prev-x86_64-pc-linux-musl/libstdc++-v3/libsupc++/.libs -fcf-protection -fPIC -c  -DIN_GCC_FRONTEND -march=native -mtune=native -m64  -fPIC -Os -fno-checking -gtoggle -DIN_GCC    -fno-exceptions -fno-rtti -fasynchronous-unwind-tables -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wmissing-format-attribute -Wconditionally-supported -Woverloaded-virtual -pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings   -DHAVE_CONFIG_H -fPIC -I. -Ijit -I/usr/src/gcc-14.1.0/gcc -I/usr/src/gcc-14.1.0/gcc/jit -I/usr/src/gcc-14.1.0/gcc/../include  -I/usr/src/gcc-14.1.0/gcc/../libcpp/include -I/usr/src/gcc-14.1.0/gcc/../libcody -I//include -I//include -I/usr/include  -I/usr/src/gcc-14.1.0/gcc/../libdecnumber -I/usr/src/gcc-14.1.0/gcc/../libdecnumber/bid -I../libdecnumber -I/usr/src/gcc-14.1.0/gcc/../libbacktrace -I/usr/src/gcc-14.1.0.bld/./isl/include -I/usr/src/gcc-14.1.0/isl/include  -o jit/jit-recording.o -MT jit/jit-recording.o -MMD -MP -MF jit/.deps/jit-recording.TPo /usr/src/gcc-14.1.0/gcc/jit/jit-recording.cc
	In file included from /usr/include/pthread.h:30,
	                 from /usr/src/gcc-14.1.0.bld/prev-x86_64-pc-linux-musl/libstdc++-v3/include/x86_64-pc-linux-musl/bits/gthr-default.h:35,
	                 from /usr/src/gcc-14.1.0.bld/prev-x86_64-pc-linux-musl/libstdc++-v3/include/x86_64-pc-linux-musl/bits/gthr.h:157,
	                 from /usr/src/gcc-14.1.0/libstdc++-v3/include/ext/atomicity.h:35,
	                 from /usr/src/gcc-14.1.0/libstdc++-v3/include/bits/ios_base.h:39,
	                 from /usr/src/gcc-14.1.0/libstdc++-v3/include/std/ios:44,
	                 from /usr/src/gcc-14.1.0/libstdc++-v3/include/std/istream:40,
	                 from /usr/src/gcc-14.1.0/libstdc++-v3/include/std/sstream:40,
	                 from /usr/src/gcc-14.1.0/gcc/jit/jit-recording.cc:32:
	/usr/include/sched.h:84:7: error: attempt to use poisoned "calloc"
	   84 | void *calloc(size_t, size_t);
	      |       ^
	In file included from /usr/src/gcc-14.1.0/gcc/jit/jit-recording.cc:22:
	/usr/src/gcc-14.1.0/gcc/system.h:938:21: note: poisoned here
	  938 |  #pragma GCC poison calloc strdup strndup
	      |                     ^~~~~~
	/usr/include/sched.h:124:36: error: attempt to use poisoned "calloc"
	  124 | #define CPU_ALLOC(n) ((cpu_set_t *)calloc(1,CPU_ALLOC_SIZE(n)))
	      |                                    ^
	/usr/src/gcc-14.1.0/gcc/system.h:938:21: note: poisoned here
	  938 |  #pragma GCC poison calloc strdup strndup
	      |                     ^~~~~~

Origin: Upstream
Upstream-Status: Submitted
 [https://gcc.gnu.org/bugzilla/show_bug.cgi?id=115442]
Signed-off-by: Ismael Luceno <ismael@sourcemage.org>
---
 gcc/jit/jit-recording.cc |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gcc/jit/jit-recording.cc b/gcc/jit/jit-recording.cc
index 68a2e860c1f..70830e34965 100644
--- a/gcc/jit/jit-recording.cc
+++ b/gcc/jit/jit-recording.cc
@@ -19,6 +19,7 @@ along with GCC; see the file COPYING3.  If not see
 <http://www.gnu.org/licenses/>.  */
 
 #include "config.h"
+#define INCLUDE_SSTREAM
 #include "system.h"
 #include "coretypes.h"
 #include "tm.h"
@@ -29,7 +30,6 @@ along with GCC; see the file COPYING3.  If not see
 #include "jit-builtins.h"
 #include "jit-recording.h"
 #include "jit-playback.h"
-#include <sstream>
 
 namespace gcc {
 namespace jit {
